### ISSUE TITLE
An issue with GRUD

### ISSUE THREAD

**Issue by JHLiu7** [MLforHealth/MIMIC_Extract#56]
Hi there, thanks for the great work in building the data pipeline!

There seems to be an issue with the implementation of GRUD in this [line](https://github.com/MLforHealth/MIMIC_Extract/blob/420bfc4122cd267ae92c2934fb4f49dcc2aed577/notebooks/mmd_grud_utils.py#L157), where `r` was not included as `combined_r`, as shown in the adapted [implementation](https://github.com/zhiyongc/GRU-D/blob/d070b52523db0dff7b6abb0a7f17ecb9baaadd42/GRUD.py#L126).

**Comment by yfeng24816** [comment]
Hello,

Are you also trying to reproduce this MIMIC_Extract thing? I was running the ipython notebook for GRU-D and ran into an error that said the training loss is NaN for the whole time.

**Comment by 2533245542** [comment]
Hello,

Are you also trying to reproduce this MIMIC_Extract thing? I was running the ipython notebook for GRU-D and ran into an error that said the training loss is NaN for the whole time.

**Comment by JHLiu7** [comment]
I made some adjustments based on this [repo](https://github.com/zhiyongc/GRU-D) and was able to train on my own dataset, though I'm not 100% sure if it has reached the expected performance. See [here](https://gist.github.com/JHLiu7/c8dda5322035c369d6433ee5ecb59d31), and let me know if there's an issue. But I haven't tried with their notebook yet. Also, here is another [implementation](https://github.com/YuliaRubanova/latent_ode/blob/master/lib/rnn_baselines.py) of gru-d that might help, but it may need more refactoring.

**Comment by 2533245542** [comment]
Much appreciated! Does your dataset contain NA values? When I was testing the MIMIC_Extract GRU-D notebook, I found `X_mean` was manually injected with NaN following this line `np.where((X_train.loc[:, pd.IndexSlice[:, 'mask']] == 1).values, 1, np.NaN)`.

The NaN makes the linear layer of the GRU-D model produced all NA results and thus not able to proceed.

Meanwhile I was trying the code from (https://github.com/zhiyongc/GRU-D) and unfortunately the data the author provided did not have NA and thus I could not see how GRU-D handled NA in a real case. If I remembered it correctly, NA handling was one of the key feature of GRU-D?

**Comment by JHLiu7** [comment]
Ah, that's right. I recall the same issue with X_mean and I ended up replacing na in X_mean by 0 as these are the measurements that were never observed in the whole training data, which may not typically be the case in a real-world scenario. gru-d indeed handles na, but I think it refers to the missing values of individual samples/time-series, so I guess it may still need to be initialized with a valid X_mean.

**Comment by bnestor** [comment]
Hi, yeah when I apply this model in other projects (https://github.com/MLforHealth/MIMIC_Generalisation/blob/master/utils/AUC_GH.py#L216), I usually do a standard scaler on the covariates (so that all the means are a tensor of zeros) then use pandas to apply `fillna(0)` to the mask tokens. Is there a PR for the grud bug?
